File: backend/app/batch.py
"""Batch delivery vehicle: `uv run python -m app.batch --ids 1-125`.

Criteria G5 asks for the whole Science-125 set with resumption. That is a loop around
the existing composition root, not a second pipeline: every question goes through
`cli.run_cli`, so a batch run and a single run produce byte-identical artifacts.

Four properties make a many-hour production run survivable:
resumption (a question whose settled run already exited 0 is never paid for twice),
isolation (one question's outage is recorded and the batch continues),
serialization (Bailian rate-limits, so questions run one at a time, in a fixed order),
and a deadline (a question that hangs is cancelled instead of stalling the night).
"""

from __future__ import annotations

import argparse
import asyncio
import json
import re
import time
from collections.abc import Sequence
from dataclasses import dataclass
from datetime import UTC, datetime
from pathlib import Path
from typing import Literal

from app.cli import run_cli
from app.domain.runs import is_run_id, render_failed, replace_text
from app.domain.science125 import find_question
from app.services.launch import RUN_TIMEOUT_SECONDS, science125_text

Status = Literal["passed", "failed", "skipped", "error", "missing", "planned"]

CLEAN = ("passed", "skipped", "planned")
"""Statuses that leave the batch's exit code at 0; everything else is a question owed."""

SAME_CLASS_STOP = 5
"""A batch is worth running because questions are independent; a same-class streak refutes that."""

OUTAGE_STOP = 2
"""Two outages in a row is credentials, network, or quota — the other 123 would fail identically."""

OUTAGE_CLASSES = frozenset({"infra_error", "infra_timeout"})

CANCEL_GRACE_SECONDS = 30.0
"""How long a cancelled question gets to unwind before the batch stops waiting for it.

`launch.RunLauncher` can `kill(2)` its hung child; a batch awaits the pipeline in its own
process and can only cancel it. A coroutine that swallows the cancellation would hang the
batch just as badly as the hang it was meant to cure, so the wait is bounded too.
"""

_RANGE = re.compile(r"(\d+)-(\d+)")
_SINGLE = re.compile(r"\d+")


@dataclass(frozen=True)
class QuestionOutcome:
    question_id: int
    status: Status
    seconds: float
    detail: str = ""
    classification: str | None = None


def parse_ids(spec: str) -> list[int]:
    """`61`, `3,54,61`, `1-125`, or any mixture; the result is deduplicated and ascending."""
    ids: set[int] = set()
    for piece in (chunk.strip() for chunk in spec.split(",")):
        if not piece:
            continue
        span = _RANGE.fullmatch(piece)
        if span is not None:
            low, high = int(span.group(1)), int(span.group(2))
            if low > high:
                raise ValueError(f"题号区间 {piece!r} 的起点大于终点。")
            ids.update(range(low, high + 1))
        elif _SINGLE.fullmatch(piece):
            ids.add(int(piece))
        else:
            raise ValueError(f"无法解析的题号片段 {piece!r}；只接受 `61`、`3,54,61` 或 `1-125`。")
    if not ids:
        raise ValueError("--ids 没有给出任何题号。")
    return sorted(ids)


def passed_question_runs(runs_root: Path) -> dict[int, str]:
    """Question id -> newest run that carries it and finished deliverable.

    `meta.questionId` ties a run to a question; a second artifact makes it terminal.
    Two shapes count as terminal, because the committed corpus contains both:

    - current: `exit.json` with exit code 0;
    - legacy: no `exit.json` (it predates that convention) but a verifier verdict of
      ALL PASS — the TypeScript-era `meta` + ALL PASS contract.

    Reading legacy runs rather than backfilling them keeps `runs/` immutable after
    terminal state, which is the whole reason it is fact data and not cache.
    """
    passed: dict[int, str] = {}
    for entry in _run_dirs(runs_root):
        question_id = _int(_read_mapping(entry / "meta.json").get("questionId"))
        if question_id is not None and _is_deliverable(entry):
            passed[question_id] = entry.name
    return passed


def _run_dirs(runs_root: Path) -> list[Path]:
    try:
        return sorted(entry for entry in runs_root.iterdir() if entry.is_dir() and is_run_id(entry.name))
    except OSError:
        return []


def _failure_class(runs_root: Path, question_id: int) -> str | None:
    """Read back the reason the Harness recorded — the only thing that can tell the batch to stop."""
    runs = [
        run for run in _run_dirs(runs_root) if _int(_read_mapping(run / "meta.json").get("questionId")) == question_id
    ]
    classification = _read_mapping(runs[-1] / "exit.json").get("classification") if runs else None
    return classification if isinstance(classification, str) and classification else None


def _is_deliverable(run: Path) -> bool:
    exit_code = _int(_read_mapping(run / "exit.json").get("exitCode"))
    if exit_code is not None:
        return exit_code == 0
    return _read_mapping(run / "verification.json").get("ok") is True or _all_pass(run)


def _all_pass(run: Path) -> bool:
    """The legacy verdict lives only in the rendered report; runs before verification.json have no other."""
    try:
        return "结果: ALL PASS" in (run / "verification-report.md").read_text(encoding="utf-8")
    except (OSError, UnicodeDecodeError):
        return False


async def run_batch(
    question_ids: Sequence[int],
    repo_root: Path,
    *,
    runs_root: Path | None = None,
    dry_run: bool = False,
    memory: bool = True,
    timeout_seconds: float = RUN_TIMEOUT_SECONDS,
) -> list[QuestionOutcome]:
    root = repo_root.resolve()
    runs = runs_root or root / "runs"
    already = passed_question_runs(runs)
    outcomes: list[QuestionOutcome] = []
    cause, streak = "", 0
    for index, question_id in enumerate(question_ids, start=1):
        outcome = await _run_one(
            question_id, root, already, runs, dry_run=dry_run, memory=memory, timeout_seconds=timeout_seconds
        )
        print(
            f"[batch] {index}/{len(question_ids)} q{outcome.question_id} | {outcome.status} | "
            f"{outcome.seconds:.1f}s{f' | {outcome.detail}' if outcome.detail else ''}",
            flush=True,
        )
        outcomes.append(outcome)
        if outcome.status == "passed":
            cause, streak = "", 0
        if outcome.status in CLEAN:
            continue
        current = outcome.classification or outcome.status
        streak = streak + 1 if current == cause else 1
        cause = current
        if streak >= (OUTAGE_STOP if current in OUTAGE_CLASSES else SAME_CLASS_STOP):
            remaining = ",".join(str(owed) for owed in question_ids[index:])
            print(
                f"[batch] 熔断停批：连续 {streak} 次 {current}。已完成 {index}/{len(question_ids)}，"
                f"剩余 --ids {remaining or '（无）'}"
            )
            break
    print(f"[batch] 合计 {len(outcomes)} 题：" + "，".join(f"{name} {count}" for name, count in _tally(outcomes)))
    return outcomes


async def _run_one(
    question_id: int,
    root: Path,
    already: dict[int, str],
    runs_root: Path,
    *,
    dry_run: bool,
    memory: bool,
    timeout_seconds: float = RUN_TIMEOUT_SECONDS,
) -> QuestionOutcome:
    if question_id in already:
        return QuestionOutcome(question_id, "skipped", 0.0, f"已有终态 passed 的 run {already[question_id]}")
    question = find_question(question_id)
    if question is None:
        return QuestionOutcome(question_id, "missing", 0.0, "题号不在 science125.json 内")
    if dry_run:
        return QuestionOutcome(question_id, "planned", 0.0, question.question[:60])
    started = time.monotonic()
    # The deadline is `launch.RUN_TIMEOUT_SECONDS` itself, not a second number: a pipeline
    # that has not settled in 40 minutes is hung whoever started it. `asyncio.wait` rather
    # than `wait_for`, because the batch must also survive the pathological case where the
    # cancellation itself does not land — `wait_for` would then wait for it forever.
    attempt = asyncio.ensure_future(run_cli(science125_text(question), root, question_id=question_id, memory=memory))
    settled, _ = await asyncio.wait({attempt}, timeout=timeout_seconds)
    if not settled:
        attempt.cancel()
        await asyncio.wait({attempt}, timeout=CANCEL_GRACE_SECONDS)
        detail = f"单题超过 {timeout_seconds:g}s 未终态，已取消"
        stuck = (
            ()
            if attempt.done()
            else (f"取消未在 {CANCEL_GRACE_SECONDS:g}s 内完成，该题可能仍在写这个 run 或持有 runs/.active.json",)
        )
        _settle_timeout(runs_root, question_id, (detail, *stuck))
        return QuestionOutcome(question_id, "failed", time.monotonic() - started, detail, "infra_timeout")
    failure = attempt.exception()
    if failure is not None:  # One question's outage must not cost the other 124.
        if not isinstance(failure, Exception):
            raise failure  # KeyboardInterrupt and friends end the batch, as they always did.
        detail = f"{type(failure).__name__}: {failure}"
        return QuestionOutcome(question_id, "error", time.monotonic() - started, detail, "infra_error")
    code = attempt.result()
    elapsed = time.monotonic() - started
    if code == 0:
        return QuestionOutcome(question_id, "passed", elapsed)
    # exit 2 is a refusal to start (missing credentials, lock held) rather than a verdict;
    # printing the code is what tells a half-hour batch log apart from a five-second one, and
    # it settles no run to read a class off, so the environment is named here instead.
    classification = _failure_class(runs_root, question_id) or ("infra_error" if code == 2 else None)
    return QuestionOutcome(question_id, "failed", elapsed, f"app.cli exit {code}", classification)


def _settle_timeout(runs_root: Path, question_id: int, failures: Sequence[str]) -> None:
    """Give the cancelled question the same terminal evidence the HTTP timeout path writes.

    A cancelled `run_cli` unwinds through its own `finally` and releases `runs/.active.json`,
    but it never reaches `_write_cli_complete`: the run dir is left with a `meta.json` and no
    verdict. Nobody else will ever come back to it, so the batch — the only surviving witness —
    settles it here, in the shape `RunLauncher._complete` uses: merge into `exit.json`, mirror
    into `meta.json`, render `FAILED.md`. Merging matters because the child already wrote facts
    (`sourceIdentity` above all) that a from-scratch file would silently drop.
    """
    run = _unsettled_run(runs_root, question_id)
    if run is None:  # The question never got as far as reserving a run dir.
        return
    finished = _now()
    exit_fact = _read_mapping(run / "exit.json")
    exit_fact.update({"exitCode": -1, "endedAt": finished, "classification": "infra_timeout"})
    replace_text(run / "exit.json", json.dumps(exit_fact, ensure_ascii=False, indent=2) + "\n")
    meta = _read_mapping(run / "meta.json")
    meta.update({"finishedAt": finished, "exitCode": -1})
    replace_text(run / "meta.json", json.dumps(meta, ensure_ascii=False, indent=2) + "\n")
    if not (run / "FAILED.md").exists():
        replace_text(run / "FAILED.md", render_failed(failures, "infra_timeout"))


def _unsettled_run(runs_root: Path, question_id: int) -> Path | None:
    """The newest run carrying this question, and only while it has no verdict of its own.

    `run_cli` writes `meta.questionId` before the pipeline starts, which is what makes the
    interrupted run findable at all. A run that already settled is somebody else's fact and
    `runs/` is immutable after terminal state, so it is never rewritten from here.
    """
    for run in reversed(_run_dirs(runs_root)):
        if _int(_read_mapping(run / "meta.json").get("questionId")) != question_id:
            continue
        return None if _int(_read_mapping(run / "exit.json").get("exitCode")) is not None else run
    return None


def _now() -> str:
    return datetime.now(UTC).isoformat(timespec="milliseconds").replace("+00:00", "Z")


def _tally(outcomes: Sequence[QuestionOutcome]) -> list[tuple[str, int]]:
    """A histogram over (status, classification): after 125 questions it is what says what to fix."""
    counts: dict[str, int] = {}
    for outcome in outcomes:
        key = outcome.status if outcome.classification is None else f"{outcome.status}/{outcome.classification}"
        counts[key] = counts.get(key, 0) + 1
    return sorted(counts.items())


def _read_mapping(path: Path) -> dict[str, object]:
    try:
        raw = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, UnicodeDecodeError, json.JSONDecodeError):
        return {}
    return raw if isinstance(raw, dict) else {}


def _int(value: object) -> int | None:
    return value if isinstance(value, int) and not isinstance(value, bool) else None


def _parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description="Run a list of Science-125 questions serially, skipping successes.")
    parser.add_argument("--ids", required=True, help="题号，例如 `1-125`、`3,54,61`，或两者混写。")
    parser.add_argument("--repo-root", type=Path, default=Path.cwd().parent, help="含 runs/ 与 memory/ 的仓库根。")
    parser.add_argument("--runs-root", type=Path, help="断点续跑扫描的 runs 目录；默认取 <repo-root>/runs。")
    parser.add_argument("--dry-run", action="store_true", help="只打印计划，一次运行都不发起。")
    parser.add_argument("--no-memory", action="store_true", help="消融臂：整批关闭跨 run 记忆，透传给 app.cli。")
    return parser


def main(argv: list[str] | None = None) -> int:
    args = _parser().parse_args(argv)
    try:
        question_ids = parse_ids(args.ids)
    except ValueError as exc:
        print(f"[batch] {exc}")
        return 2
    outcomes = asyncio.run(
        run_batch(
            question_ids,
            args.repo_root,
            runs_root=args.runs_root,
            dry_run=args.dry_run,
            memory=not args.no_memory,
        )
    )
    return 0 if all(outcome.status in CLEAN for outcome in outcomes) else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: frontend/src/batch.ts
/**
 * 批次概览的全部计算：runs 列表 + Science-125 题库 -> 一页能读的概览。
 *
 * 这里没有任何新的运行时状态。125 批跑的事实早已躺在 `runs/<id>/` 里——`meta.questionId`
 * 说这次跑的是哪一题，`exit.json` 说它怎么结束的、由哪份代码产出——本模块只是换个视角
 * 把它们聚起来。所以它是纯函数：同一份输入永远得到同一份输出，可以直接单测。
 *
 * 语义的权威在后端，不在这里：
 * - 失败分类的取值域是 `backend/app/domain/runs.py` 的 `FailureClass`；
 * - 环境性/质量性的切分是 `backend/app/evaluation.py` 的 `INFRASTRUCTURE_CLASSES`；
 * - cohort 标签的写法对齐 `evaluation.py` 的 `_cohort_label`（脏树是标签的一部分）；
 * - 题号区间的写法对齐 `backend/app/batch.py` 的 `parse_ids`（`3,7,12-14`）。
 */

import type { RunStatus, RunSummary, Science125 } from "./types"

/**
 * 环境故障不是质量判决：arXiv 超时说明的是网络，不是这份研究计划好不好。
 * 两者混在一个数字里，「我该重跑这批还是该改代码」就没有判据了。
 */
const INFRASTRUCTURE_CLASSES = new Set(["infra_error", "infra_timeout"])

export type FailureKind = "infra" | "quality" | "unclassified"

/** 一题一行：这题最终怎么了，以及是哪次 run 说的。 */
export type QuestionOutcome = {
  questionId: number
  question: string
  domain: string | null
  /** 代表这题终态的那次 run；点它进详情。 */
  runId: string
  status: RunStatus
  classification: string | null
  kind: FailureKind | null
  /** 这题一共跑过几次——多次运行本身是信号。 */
  attempts: number
}

export type FailureGroup = {
  /** 未分类失败没有 classification，用 null 表达，不编一个 "unknown" 出来。 */
  classification: string | null
  kind: FailureKind
  questions: QuestionOutcome[]
}

export type Cohort = {
  /** git 不可用时终态写的就是 null；此时这批数字无法归因到某份代码。 */
  commit: string | null
  treeDirty: boolean
  questions: number
}

export type BatchOverview = {
  /** 题库题数，不是写死的 125。 */
  total: number
  /** 至少跑过一次的题数。 */
  attempted: number
  passed: number
  failed: number
  working: number
  /** 环境性失败的题数。 */
  infra: number
  /** 质量性失败的题数。 */
  quality: number
  /** 失败但终态没自报分类的题数（TS 时代的旧 run）。 */
  unclassified: number
  /** 一次都没跑过的题号，升序。回答「跑到哪了」。 */
  notRun: number[]
  /**
   * 批跑还欠着的题号，升序——`notRun` ∪「跑过但一次都没通过」。回答「接下来跑什么」。
   *
   * 判据与 `app.batch` 的 `passed_question_runs` 同源，且是照它的**实现**定的，不是照直觉：
   * 那个函数遍历升序的 run 目录，只在 `_is_deliverable(entry)` 为真时才写
   * `passed[question_id] = entry.name`，失败的 run 既不覆盖也不删除已有的键。所以它认的是
   * **「存在任一 passed run」**，不是「最新终态是 passed」——先通过后失败的题，batch 照样跳过，
   * 因此它不欠。这与 `failed`（按最新终态算）故意不同：一个说进度，一个说欠账。
   */
  owed: number[]
  /** 已按 kind（infra → quality → 未分类）排好序的失败题清单。 */
  failures: FailureGroup[]
  /** 代表各题终态的那批 run 的 commit 分布；长度 > 1 即这些数字不是同一个系统产生的。 */
  cohorts: Cohort[]
  /** 没有题号的 run 数（自由输入 / 题号不在题库内）；它们不进任何覆盖率。 */
  unattributed: number
}

const KIND_ORDER: Record<FailureKind, number> = {
  infra: 0,
  quality: 1,
  unclassified: 2,
}

const failureKind = (classification: string | null): FailureKind =>
  classification === null
    ? "unclassified"
    : INFRASTRUCTURE_CLASSES.has(classification)
      ? "infra"
      : "quality"

/**
 * 题号列表压成 `app.batch --ids` 认的写法。三个以上的连号才压成区间：
 * `12,13` 与 `12-13` 一样长，压了反而多一次心算。
 */
export function compactIds(ids: readonly number[]): string {
  const sorted = [...new Set(ids)].sort((a, b) => a - b)
  const parts: string[] = []
  let start = 0
  while (start < sorted.length) {
    let end = start
    while (end + 1 < sorted.length && sorted[end + 1] === sorted[end] + 1)
      end += 1
    if (end - start + 1 >= 3) parts.push(`${sorted[start]}-${sorted[end]}`)
    else for (let i = start; i <= end; i += 1) parts.push(String(sorted[i]))
    start = end + 1
  }
  return parts.join(",")
}

/**
 * run id 形如 `20260810-165229`，字典序即时间序（后端 `list_ids` 也这么排）。
 * 用它而不是 `startedAt`：回填的旧 run 会把 startedAt 写成同一秒，排不出先后。
 */
const newestFirst = (a: RunSummary, b: RunSummary) =>
  a.id < b.id ? 1 : a.id > b.id ? -1 : 0

/**
 * 一题多次运行时，代表它的是**最新的终态 run**。全是 working（只可能是那唯一活跃的
 * 一次）时退回最新的一次，这题就报 working——不能因为还没终态就当它没跑过。
 */
const representative = (ordered: RunSummary[]): RunSummary =>
  ordered.find((run) => run.status !== "working") ?? ordered[0]

export function summarizeBatch(
  runs: readonly RunSummary[],
  science: Science125,
): BatchOverview {
  const bank = new Map<number, { question: string; domain: string }>()
  for (const domain of science.domains)
    for (const question of domain.questions)
      bank.set(question.id, {
        question: question.question,
        domain: domain.domain,
      })

  const byQuestion = new Map<number, RunSummary[]>()
  let unattributed = 0
  for (const run of runs) {
    const questionId = run.science125Id
    // 题号不在题库里的 run（自由输入的 OOD run）无法进覆盖率的分母，只能单列。
    if (questionId === null || !bank.has(questionId)) {
      unattributed += 1
      continue
    }
    const existing = byQuestion.get(questionId)
    if (existing) existing.push(run)
    else byQuestion.set(questionId, [run])
  }

  // 遍历题库而不是遍历 run：分母是题库，「一次没跑过」才有地方被数出来。
  const settled: { outcome: QuestionOutcome; run: RunSummary }[] = []
  const notRun: number[] = []
  const owed: number[] = []
  for (const [questionId, entry] of bank) {
    const group = byQuestion.get(questionId)
    if (group === undefined) {
      notRun.push(questionId)
      owed.push(questionId)
      continue
    }
    // 「存在任一 passed run」——照 `passed_question_runs` 的实现，不是按最新终态。
    if (!group.some((candidate) => candidate.status === "passed"))
      owed.push(questionId)
    const run = representative([...group].sort(newestFirst))
    const classification = run.status === "failed" ? run.classification : null
    settled.push({
      run,
      outcome: {
        questionId,
        question: entry.question,
        domain: entry.domain,
        runId: run.id,
        status: run.status,
        classification,
        kind: run.status === "failed" ? failureKind(classification) : null,
        attempts: group.length,
      },
    })
  }
  settled.sort((a, b) => a.outcome.questionId - b.outcome.questionId)
  notRun.sort((a, b) => a - b)
  owed.sort((a, b) => a - b)

  const outcomes = settled.map((item) => item.outcome)
  const failed = outcomes.filter((outcome) => outcome.status === "failed")
  const counted = (kind: FailureKind) =>
    failed.filter((outcome) => outcome.kind === kind).length

  return {
    total: bank.size,
    attempted: outcomes.length,
    passed: outcomes.filter((outcome) => outcome.status === "passed").length,
    failed: failed.length,
    working: outcomes.filter((outcome) => outcome.status === "working").length,
    infra: counted("infra"),
    quality: counted("quality"),
    unclassified: counted("unclassified"),
    notRun,
    owed,
    failures: groupFailures(failed),
    cohorts: cohortsOf(settled.map((item) => item.run)),
    unattributed,
  }
}

/** 环境性在前、质量性在后，同类里题多的在前——先看该查凭据还是该改代码。 */
function groupFailures(failed: readonly QuestionOutcome[]): FailureGroup[] {
  const groups = new Map<string, FailureGroup>()
  for (const outcome of failed) {
    const kind = outcome.kind ?? "unclassified"
    const key = outcome.classification ?? ""
    const existing = groups.get(key)
    if (existing) existing.questions.push(outcome)
    else
      groups.set(key, {
        classification: outcome.classification,
        kind,
        questions: [outcome],
      })
  }
  return [...groups.values()].sort(
    (a, b) =>
      KIND_ORDER[a.kind] - KIND_ORDER[b.kind] ||
      b.questions.length - a.questions.length ||
      (a.classification ?? "").localeCompare(b.classification ?? ""),
  )
}

/**
 * 统计的是**代表各题终态的那批 run**的身份，而不是全部 run：页面上的数字由它们产生，
 * 该被追问「是同一份代码吗」的也只有它们。
 */
function cohortsOf(runs: readonly RunSummary[]): Cohort[] {
  const cohorts = new Map<string, Cohort>()
  for (const run of runs) {
    const commit = run.sourceIdentity?.gitCommit ?? null
    const treeDirty = run.sourceIdentity?.treeDirty === true
    const key = commit === null ? "" : treeDirty ? `${commit}+dirty` : commit
    const existing = cohorts.get(key)
    if (existing) existing.questions += 1
    else cohorts.set(key, { commit, treeDirty, questions: 1 })
  }
  return [...cohorts.values()].sort(
    (a, b) =>
      b.questions - a.questions ||
      (a.commit ?? "").localeCompare(b.commit ?? ""),
  )
}
